File: go/reach.go
/* Copyright (c) 2025 Richard Rodger, MIT License */

package aontu


import (
	"sort"
	"strconv"
	"strings"
)

// ReachVerdict is reaches | unreachable | error.
type ReachVerdict = string

// ReachReport is the answer for one document, mirroring ReachReport in
// ts/src/reach.ts field for field.
type ReachReport struct {
	Verdict ReachVerdict `json:"verdict"`

	Path []string `json:"path,omitempty"`

	// Errors is WHY the graph could not be looked at, in vet's finding
	// shape. Present ONLY on `error`.
	Errors []VetFinding `json:"errors,omitempty"`
}

// ReachOptions mirrors ReachOptions in ts/src/reach.ts.
type ReachOptions struct {
	Relation string
}

func reachEndpointFinding(name string, known []string) VetFinding {
	f := VetFinding{
		Code:     "refer_unresolved",
		Class:    "reference",
		Severity: "error",
		Path:     "$",
		Message:  name + " names no node in this document.",
		Sites:    []VetSite{},
	}
	if 0 < len(known) {
		note := "nodes with links: " + strings.Join(known, ", ")
		f.Note = &note
	}
	return f
}

func parseNodePath(s string) ([]string, bool) {
	if "$" == s {
		return nil, true
	}
	if !strings.HasPrefix(s, "$.") {
		return nil, false
	}
	parts := strings.Split(s[2:], ".")
	for _, p := range parts {
		if !addrSegmentOK(p) {
			return nil, false
		}
	}
	return parts, true
}

func nodeAt(root Val, path []string) bool {
	node := root
	for _, seg := range path {
		var next Val
		switch n := node.(type) {
		case *MapVal:
			next = n.peg[seg]
		case *ListVal:
			if i, err := strconv.Atoi(seg); nil == err && 0 <= i && i < len(n.peg) {
				next = n.peg[i]
			}
		}
		if nil == next {
			return false
		}
		node = next
	}
	return nil != node
}

// Reach answers whether `to` is reachable from `from` over the link
// graph of src. Endpoints are `$.dotted` node paths (ADR-014).
func (a *Aontu) Reach(src, from, to string, opts *ReachOptions) ReachReport {
	options := ReachOptions{}
	if nil != opts {
		options = *opts
	}

	parsed, perr := a.parseEntry(src)
	if nil != perr {
		return ReachReport{Verdict: "error",
			Errors: []VetFinding{parseFinding(a.File, VetRoleData, perr)}}
	}

	root, ctx, _ := a.unifyCtx(parsed, nil, src)
	// A document that does not stand up is not a document with an
	// unreachable pair: the errors it already has are the answer.
	if nil == root || root.Nil() || 0 < len(ctx.err) {
		return ReachReport{Verdict: "error",
			Errors: []VetFinding{failureFinding(ctx, a.File, src, root)}}
	}

	graph := GraphOf(root)
	seen := map[string]bool{}
	linked := []string{}
	for _, e := range graph.Edges {
		for _, p := range []string{e.From, e.To} {
			if !seen[p] {
				seen[p] = true
				linked = append(linked, p)
			}
		}
	}
	sort.Strings(linked)

	errs := []VetFinding{}
	for _, n := range []string{from, to} {
		parts, pok := parseNodePath(n)
		if !pok || !nodeAt(root, parts) {
			errs = append(errs, reachEndpointFinding(n, linked))
		}
	}
	if 0 < len(errs) {
		return ReachReport{Verdict: "error", Errors: errs}
	}

	// The successor map, restricted to one relation when the caller
	// asked for one. Sorted, so the path the search finds is the same
	// one in both ports.
	succ := map[string][]string{}
	for _, e := range graph.Edges {
		if "" != options.Relation && options.Relation != e.Key {
			continue
		}
		dest := e.To
		dup := false
		for _, d := range succ[e.From] {
			if d == dest {
				dup = true
				break
			}
		}
		if !dup {
			succ[e.From] = append(succ[e.From], dest)
		}
	}
	for _, list := range succ {
		sort.Strings(list)
	}

	// BREADTH-FIRST, so the path reported is a SHORTEST one, and with
	// the successors sorted, a determined one.
	prev := map[string]string{}
	visited := map[string]bool{}
	front := []string{from}
	for 0 < len(front) {
		next := []string{}
		for _, node := range front {
			for _, dest := range succ[node] {
				if dest == to {
					path := []string{dest}
					step := node
					for step != from {
						path = append([]string{step}, path...)
						step = prev[step]
					}
					path = append([]string{from}, path...)
					return ReachReport{Verdict: "reaches", Path: path}
				}
				if !visited[dest] {
					visited[dest] = true
					prev[dest] = node
					next = append(next, dest)
				}
			}
		}
		front = next
	}

	return ReachReport{Verdict: "unreachable"}
}

File: ts/src/reach.ts
/* Copyright (c) 2025 Richard Rodger, MIT License */
import { includeOpts } from './utility'


import { Aontu } from './aontu'
import { failureFinding } from './vet'
import type { VetFinding } from './vet'
import type { TrustOptions } from './type'
import { graphOf } from './graph'
import { cmpCodePoint } from './keyorder'


export type ReachVerdict = 'reaches' | 'unreachable' | 'error'

export type ReachReport = {
  verdict: ReachVerdict

  path?: string[]

  errors?: VetFinding[]
}

export type ReachOptions = {
  // Where the document CAME FROM, so a relative `@"file"` load inside
  // it resolves from its own directory (relationCheck's precedent).
  path?: string
  // The include capability this document evaluates under (G5,
  // docs/trust.md).
  trust?: TrustOptions

  textExt?: string[]
  // Follow only edges under this relation. Absent means follow every
  // edge, which is the whole graph and the commoner question.
  relation?: string
}


export function parseNodePath(s: string): string[] | undefined {
  if ('$' === s) {
    return []
  }
  if (!s.startsWith('$.')) {
    return undefined
  }
  const parts = s.slice(2).split('.')
  return parts.every((p) => /^[A-Za-z0-9_-]+$/.test(p)) ? parts : undefined
}


function nodeAt(root: any, path: string[]): boolean {
  let node: any = root
  for (const seg of path) {
    if (true !== node?.isMap && true !== node?.isList) {
      return false
    }
    node = node.peg[seg]
    if (null == node) {
      return false
    }
  }
  return null != node
}


function endpointFinding(name: string, known: string[]): VetFinding {
  return {
    code: 'refer_unresolved',
    class: 'reference',
    severity: 'error',
    path: '$',
    message: `${name} names no node in this document.`,
    sites: [],
    ...(0 === known.length ? {} : {
      note: 'nodes with links: ' + known.join(', '),
    }),
  }
}


// The reachability check for one document.
export function reachCheck(
  src: string, from: string, to: string, opts?: ReachOptions
): ReachReport {
  const options = opts ?? {}
  const aontu = new Aontu(includeOpts(options))
  const ctx = aontu.ctx({ collect: true })
  const parseOpts = null == options.path ? undefined : { path: options.path }
  const root: any = aontu.unify(src, parseOpts, ctx)

  // A document that does not stand up is not a document with an
  // unreachable pair: the errors it already has are the answer.
  if (0 < ctx.err.length || true === root?.isNil) {
    return {
      verdict: 'error',
      errors: [failureFinding(ctx, options.path, root)],
    }
  }

  const graph = graphOf(root)
  // The nodes the graph actually touches, for the error note: a
  // document has every path in it, and listing them all would drown the
  // one fact a mistyped endpoint needs.
  const linked = [...new Set(graph.edges
    .flatMap((e) => [e.from, e.to]))].sort(cmpCodePoint)
  const missing = [from, to].filter((n) => {
    const parts = parseNodePath(n)
    return undefined === parts || !nodeAt(root, parts)
  })
  if (0 < missing.length) {
    return {
      verdict: 'error',
      errors: missing.map((n) => endpointFinding(n, linked)),
    }
  }

  // The successor map, restricted to one relation when the caller asked
  // for one. Sorted, so the path the search finds is the same one in
  // both ports.
  const succ = new Map<string, string[]>()
  for (const e of graph.edges) {
    if (null != options.relation && options.relation !== e.key) {
      continue
    }
    const list = succ.get(e.from)
    const dest = e.to
    if (undefined === list) {
      succ.set(e.from, [dest])
    }
    else if (!list.includes(dest)) {
      list.push(dest)
    }
  }
  for (const list of succ.values()) {
    list.sort(cmpCodePoint)
  }

  const prev = new Map<string, string>()
  const seen = new Set<string>()
  let front: string[] = [from]
  while (0 < front.length) {
    const next: string[] = []
    for (const node of front) {
      for (const dest of succ.get(node) ?? []) {
        if (dest === to) {
          const path = [dest]
          let step = node
          while (step !== from) {
            path.unshift(step)
            step = prev.get(step) as string
          }
          path.unshift(from)
          return { verdict: 'reaches', path }
        }
        if (!seen.has(dest)) {
          seen.add(dest)
          prev.set(dest, node)
          next.push(dest)
        }
      }
    }
    front = next
  }

  return { verdict: 'unreachable' }
}
